### DISCUSSION TITLE
models being mut (forward)

### DISCUSSION THREAD

**Discussion by jondot** [EricLBuehler/mistral.rs#482] [Q&A]
Hi, I've been studying the code, and wondering about the approach of the models being `mut` which cause the pipelines to be `mut`. Looking at that, indeed they're arc/mutex protected, however -- is there any state that lingers between unrelated requests when the server is running?

**Comment by EricLBuehler** [comment]
Hi @jondot! You are correct that the forward methods of the *models* do not need to be `&mut` - there is no shared state across forward passes other than the KV cache, which is managed by the [CacheManager](https://github.com/EricLBuehler/mistral.rs/blob/333ce88493f2500da016488a56f60f96831c459e/mistralrs-core/src/pipeline/cache_manager.rs#L162) and handled safely.

However, `Pipeline::step` cannot take `&self` due to `NormalModel::activate_adapters`/`IsqPipelineMixin::re_isq_model` and others which take `&mut self`. Changes made here are persistent across requests because they affect the model itself.

I'll make some changes to remove the unnecessary `&mut`s, thank you for pointing this out. If you have any other questions, please feel free to raise an issue or contact me on our Discord server!

**Comment by EricLBuehler** [comment]
@jondot I converted this to a discussion to continue our conversation, please feel free to open another issue!
